apps/nextjs/src/app/api/aila-download-all/route.ts

```diff
@@ -0,0 +1,212 @@
+import { LessonExportType, prisma } from "@oakai/db";
+import { downloadDriveFile } from "@oakai/exports";
+import * as Sentry from "@sentry/node";
+import archiver from "archiver";
+import { PassThrough } from "stream";
+
+import { withSentry } from "@/lib/sentry/withSentry";
+
+function getReadableExportType(exportType: LessonExportType) {
+  switch (exportType) {
+    case "EXIT_QUIZ_DOC":
+      return "Exit quiz";
+    case "LESSON_PLAN_DOC":
+      return "Lesson plan";
+    case "STARTER_QUIZ_DOC":
+      return "Starter quiz";
+    case "WORKSHEET_SLIDES":
+      return "Worksheet";
+    case "LESSON_SLIDES_SLIDES":
+      return "Lesson slides";
+    case "ADDITIONAL_MATERIALS_DOCS":
+      return "Additional materials";
+  }
+}
+
+async function saveDownloadEvent({
+  lessonExportId,
+  downloadedBy,
+  ext,
+}: {
+  lessonExportId: string;
+  downloadedBy: string;
+  ext: string;
+}) {
+  try {
+    await prisma.lessonExportDownload.create({
+      data: {
+        lessonExportId,
+        downloadedBy,
+        ext,
+      },
+    });
+  } catch (error) {
+    Sentry.captureException(error, {
+      level: "warning",
+      extra: { lessonExportId, downloadedBy, ext },
+    });
+  }
+}
+
+function nodePassThroughToReadableStream(passThrough: PassThrough) {
+  return new ReadableStream({
+    start(controller) {
+      passThrough.on("data", (chunk) => {
+        controller.enqueue(new Uint8Array(chunk));
+      });
+      passThrough.on("end", () => {
+        controller.close();
+      });
+      passThrough.on("error", (err) => {
+        controller.error(err);
+      });
+    },
+    cancel() {
+      passThrough.destroy();
+    },
+  });
+}
+
+async function getHandler(req: Request): Promise<Response> {
+  const { searchParams } = new URL(req.url);
+  const fileIdsParam = searchParams.get("fileIds");
+  const lessonTitle = searchParams.get("lessonTitle") ?? "";
+  const { userId }: { userId: string | null } = auth();
+
+  if (!userId) {
+    const error = new Error("Download attempt without userId");
+    Sentry.captureException(error, { level: "warning" });
+    return new Response("Unauthorized", { status: 401 });
+  }
+
+  if (!fileIdsParam) {
+    return new Response("Invalid or missing fileIds", { status: 400 });
+  }
+
+  const nonce = req.headers.get("x-middleware-csp-nonce");
+  if (!nonce) {
+    return new Response("Missing nonce", { status: 400 });
+  }
+
+  const prepareDownload = searchParams.get("prepareDownload");
+  if (!prepareDownload) {
+    const loadingHtml = `
+<html>
+  <body>
+    <pre>
+Loading...
+| 
+Please wait while we prepare your files for download. This can take up to 1 minute.
+
+<script nonce="${nonce}">
+  const spinnerChars = ['|', '/', '-', '\\\\'];
+  let spinnerIndex = 0;
+  
+  setInterval(() => {
+    document.querySelector('pre').textContent = 'Loading...\\n' + spinnerChars[spinnerIndex] + '\\nPlease wait while we prepare your files for download. This can take up to 1 minute.';
```

Comment:
😵‍💫

Reply:
I don't think we need this .. i think we can just have [`a href="api_download_route" download /`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download)`

Reply:
can chat about this on slack but that will lead the user to a blank page which they have to wait at for a min so they need some sort of indication that they have to wait.